app/models/ems_refresh/save_inventory_helper.rb

```diff
@@ -74,6 +91,60 @@ def save_inventory_multi(association, hashes, deletes, find_key, child_keys = []
     association.push(new_records)
   end
 
+  # Saves a inventory record, and any sub inventory records, without
+  # associating it with the parent object that is passed in.  Parent usually is
+  # an EMS, but doesn't have to be.
+  #
+  # @param parent [ActiveRecord] base record being saved
+  # @param association [Symbol]  associated record type being added to parent
+  # @param hashes [Array]        inventory data (hash per record)
+  # @param find_key [Array]      ordered keys for existing record lookup
+  # @param child_keys [Array]    keys from hashes that are child records
+  # @param extra_keys [Array]    hash keys to be removed before saving records
+  # @param disconnect [Boolean]  whether or not to delete or disconnect records
+  #                              that don't exist in the inventory hashes, but
+  #                              are present in the DB
+  #
+  # @return [Hash] the hashes passed into the method
+  def save_inventory_with_thin_association(parent, association, hashes, find_key,
+                                           child_keys = [], extra_keys = [],
+                                           disconnect = false)
+
+    relation      = parent.association(association).reflection
+    model         = parent.association(association).klass
+    assoc_attrs   = parent.association(association).reader.scope_attributes
+    deletes_index = model.pluck(relation.active_record_primary_key)
+                         .index_by { |x| x }
+
+    child_keys = Array.wrap(child_keys)
+    remove_keys = Array.wrap(extra_keys) + child_keys
+    record_index = IdTypedIndex.new(deletes_index.values, find_key, model)
+    inventory_ids = []
+
+    hashes.each do |h|
+      found = nil
+      ActiveRecord::Base.transaction do
```

Comment:
can we do hashes.slice and do transaction e.g. for each 1000 records? As we discussed with @kbrock

Reply:
I will test with that, but one advantage of doing it one by one is the garbage collection happens quicker and doesn't balloon as much.  I will try some tests around that, but I was mostly solving for memory in this effort, and not CPU (even though I got a bit of savings with this one in CPU already by coincidence).

Reply:
hm, I remember testing this and the transaction was not affecting memory. Maybe I tested it badly though. I've seen the same mem peaks with doing transaction per 1 and transaction per all though.

Reply:
I think what you would end up seeing with the memory in this loop is it would look like this when doing batches:

```
  |
m |     /\      /\      /\      /\      
e |    /  \    /  \    /  \    /  \    
m |   /    \  /    \  /    \  /    \  
  |  /      \/      \/      \/      \
  ------------------------------------
                time
```

Versus doing it individually:

```
  |
m |
e |
m |
  |  /\/\/\/\/\/\/\/\/\/\/\/\/\/\/\/\
  ------------------------------------
                time
```

(sorry for the ascii graphs...)


And unless you were capturing the memory consistently, you might not have caught the peaks if you captured after the loop and garbage collection already happened.

That said, the peaks should be far lower then what we are getting now, so it might be worth doing, though possibly as a second pass after we can get some better metrics.

Reply:
Hm and you have tried this just with moving the transactions? Transaction should not hold any references, unless in a case I stated above ^. When I tested it, i saw only the association was holding the references.

Reply:
If possible, more rows in a transaction should speed this up. (this has already been mentioned)

I'm surprised that the numbers didn't take too big of a hit from the multiple transactions.

I like the ascii graphs...